src/DynamicRupture/Output/DataTypes.hpp

```diff
@@ -0,0 +1,168 @@
+#ifndef SEISSOL_DR_OUTPUT_DATA_TYPES_HPP
+#define SEISSOL_DR_OUTPUT_DATA_TYPES_HPP
+
+#include "Geometry.hpp"
+#include "Initializer/tree/Layer.hpp"
+#include "generated_code/tensor.h"
+#include <Eigen/Dense>
+#include <array>
+#include <cassert>
+#include <cstring>
+#include <limits>
+#include <tuple>
+#include <vector>
+
+namespace seissol::dr::output {
+template <int DIM>
+struct VarT {
+  constexpr int dim() { return DIM; }
+
+  real* operator[](int dim) {
+    assert(dim < DIM && "access is out of the DIM. bounds");
+    assert(data[dim] != nullptr && "data has not been initialized yet");
+    return data[dim];
+  }
+
+  real& operator()(int dim, size_t level, size_t index) {
+    assert(dim < DIM && "access is out of DIM. bounds");
+    assert(level < maxCacheLevel && "access is out of cache bounds");
+    assert(index < size && "access is out of size bounds");
+    assert(data[dim] != nullptr && "data has not been initialized yet");
+    return data[dim][index + level * size];
+  }
+
+  real& operator()(size_t level, size_t index) {
+    static_assert(DIM == 1, "access of the overload is allowed only for 1 dim variables");
+    return this->operator()(0, level, index);
+  }
+
+  // allocates data for a var (for all dimensions and cache levels)
+  // initialized to zeros if var is active.
+  // Otherwise, inits with nullptr
+  void allocateData(size_t dataSize) {
+    size = dataSize;
+    if (isActive) {
+      for (int dim = 0; dim < DIM; ++dim) {
+        data[dim] = new real[size * maxCacheLevel];
+        std::memset(static_cast<void*>(data[dim]), 0, size * maxCacheLevel * sizeof(real));
+      }
+    } else {
+      for (int dim = 0; dim < DIM; ++dim)
+        data[dim] = nullptr;
+    }
+  }
+
+  void releaseData() {
+    if (isActive) {
+      for (auto item : data) {
+        delete[] item;
+      }
+    }
+  }
+
+  std::array<real*, DIM> data{};
+  bool isActive{false};
+  size_t size{};
+  size_t maxCacheLevel{1};
+};
+
+using Var1D = VarT<1>;
+using Var2D = VarT<2>;
+using Var3D = VarT<3>;
+
+// Description is given in `enum VariableID`
+using DrVarsT =
+    std::tuple<Var2D, Var3D, Var1D, Var2D, Var3D, Var2D, Var1D, Var1D, Var1D, Var1D, Var1D, Var2D>;
+
+enum DirectionID { Strike = 0, Dip = 1, Normal = 2 };
+enum TPID { Pressure = 0, Temperature = 1 };
+enum ParamID { FrictionCoefficient = 0, State = 1 };
+
+enum VariableID {
+  SlipRate = 0,
+  TransientTractions,
+  NormalVelocity,
+  FrictionAndState,
+  TotalTractions,
+  Slip,
+  RuptureVelocity,
+  AccumulatedSlip,
+  PeakSlipRate,
+  RuptureTime,
+  DynamicStressTime,
+  TpVariables, // Thermal Pressurization
+  Size
+};
+
+enum class OutputType : int {
+  None = 0,
+  AtPickpoint = 3,
+  Elementwise = 4,
+  AtPickpointAndElementwise = 5
+};
+
+enum class SlipRateOutputType { VelocityDifference, TractionsAndFailure };
+
+struct GeneralParamsT {
```

Comment:
Remove T at the end. The struct belong to `seissol::dr::output` namespace and it is supposed to be used internally within dr output system. I don't see a point of renaming it because it is not exposed to a user.
 
See #645